rfcs/convergence/deprecate-keyboard-key-package.md

```diff
@@ -0,0 +1,94 @@
+
+Proposes to break the dependency of any v9 packages on `@fluentui/keyboard-key` since it is not a converged package and can increase bundle size.
+Also proposes to stop using `keyCode` in v9 components and start using `key`.
+
+## Problem statement
+
+```typescript
+import { ArrowLeftKey } from '@fluentui/keyboard-key';
+
+console.log(ArrowLeftKey);
+```
+
+The above code will increase the bundle size by **2kb** minified and **1.0kb** gzipped. Looking at the output, every single key is retained in the bundle. The measurement was done using the bunde-size tool implemented as a part of [#17697](https://github.com/microsoft/fluentui/issues/17697).
+
+According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) `keyCode` has been deprecated from some time and developers should prefer to use `e.key` which takes into consideration modifier keys and the locale and layout. IE11 handles the values differently to other evergreen browsers, a few examples:
+
+| Everyone else | IE11     |
+| ------------- | -------- |
+| ArrowDown     | Down     |
+| Escape        | Esc      |
+| ' '           | Spacebar |
+
+## Detailed Design or Proposal
+
+### Remove dual v8/v9 dependency on `@fluentui/keyboard-key`
+
+Since there are v9 components now being used in HVCs and products, this might be difficult since this could result in a breaking change in terms of dependencies.
+
+### Create new package
+
+We could create a new `@fluentui/react-keyboard-key` (naming flexible from this RFC) package to store these constants. The one advantage this would have over the previous option is more readable imports:
+
+```typescript
+import { ArrowDown, Escape } from '@fuentui/react-keyboard-key';
+```
+
+instead of
+
+```typescript
+import { ArrowDown, useIsomorphicLayoutEffect, Escape, useEventCallback } from '@fuentui/react-utilities';
+```
+
+> The below statement should be verified
+
+There should be no major problem with dependencies since older packages will still work with the old keyboard-key package. We generally do not share any logic with keys in our shared contexts or cross components.
+
+### Use e.key
+
+Since we no longer target support for IE11, the usage of `e.key` is simplified and no longer needs to have special handling for IE11. We can modify an existing package or create a new one that simply stores all of the keys as constants which are easily tree shaken.
+
+A list of support for `key` [can be found in MDN](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key#browser_compatibility)
+
+### Compatibility with e.keyCode
+
+In order to comfortably use both `key` and `keyCode` during a transition phase and also in the future if we ever find out one might be preferrable to another, we should implement the new package to be compatible with both. This can be done with a `key` and `keycode` naming convention.
+
+```typescript
+import { ArrowDownKey, ArrowDownKeyCode } from '@fluentui/react-keyboard-key';
+```
+
```

Comment:
Current package (`keyboard-key`) contains `getKey()` and `getCode()` helpers for compatibility reasons. Do I correctly get that a new package will contain only constants with keys and their codes? If so, can you please add this statement in the proposal?

Reply:
yes there will be no more helpers, since the helpers helped to contribute to the lack of tree shaking for the current package.

There is really not that much work to do to call `e.key` or `e.keyCode` now that we support modern browsers